### PR TITLE
Implement key backend support

### PR DESCRIPTION
**Pull request by mathiasertl**
Implement key backend support, providing the ability for using different backends such as HSMs.

See also: #129

### CODE REVIEW

File: ca/django_ca/models.py [hunk 1]
```diff
@@ -497,7 +492,8 @@ class CertificateAuthority(X509CertMixin):
     parent = models.ForeignKey(
         "self", on_delete=models.SET_NULL, null=True, blank=True, related_name="children"
     )
-    private_key_path = models.CharField(max_length=256, help_text=_("Path to the private key."))
+    key_backend_path = models.CharField(max_length=256, help_text=_("Backend to handle private keys."))
+    key_backend_options = models.JSONField(default=dict, help_text=_("Key backend options"))
```
Comment (mathiasertl) [inline ca/django_ca/models.py:496]: `key_backend_path` is a Python class path and should represent a type implementing a certain interface.
 `key_backend_options` is the options to pass to an instance of that interface.

File: ca/django_ca/models.py [hunk 2]
```diff
@@ -604,37 +600,11 @@ class Meta:
...
-    def key_exists(self) -> bool:
-        """``True`` if the private key is accessible to the current process."""
-        if self._key is not None:
-            return True
-        return file_exists(self.private_key_path)
+    def get_key_backend(self) -> "KeyBackend":
+        module_path, class_name = self.key_backend_path.rsplit(".", 1)
+        module = importlib.import_module(module_path)
+        cls = getattr(module, class_name)
+        return cls(ca=self, **self.key_backend_options)
```
Comment (mathiasertl) [inline ca/django_ca/models.py:607]: This function allows you to get a key backend instance for the particular Certificate Authority. The instance should be capable of generating its own private key and sign certificates.

File: ca/django_ca/ca_settings.py [hunk 3]
```diff
@@ -329,13 +329,17 @@ def _get_hash_algorithm(setting: str, default: "HashAlgorithms") -> "AllowedHash
...
 
 
-# CA_DEFAULT_ECC_CURVE can be removed in django-ca==1.25.0
 _CA_DEFAULT_ELLIPTIC_CURVE = getattr(settings, "CA_DEFAULT_ELLIPTIC_CURVE", ec.SECP256R1.name)
 try:
     CA_DEFAULT_ELLIPTIC_CURVE = constants.ELLIPTIC_CURVE_TYPES[_CA_DEFAULT_ELLIPTIC_CURVE]
 except KeyError as ex:
     raise ImproperlyConfigured(f"{_CA_DEFAULT_ELLIPTIC_CURVE}: Unknown CA_DEFAULT_ELLIPTIC_CURVE.") from ex
 
+CA_KEY_BACKENDS: Tuple[str] = tuple(
```
Comment (kushaldas) [inline ca/django_ca/ca_settings.py:338]: Maybe a comment explaining what to set here? The different options/values.

File: ca/django_ca/backends/base.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        self.alias = alias
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    @property
+    def class_path(self) -> str:
+        """Shortcut returning the full Python class path of this instance."""
+        return f"{self.__class__.__module__}.{self.__class__.__name__}"
+
+    def add_private_key_group(self, parser: CommandParser) -> Optional[ArgumentGroup]:
```
Comment (mathiasertl) [inline ca/django_ca/backends/base.py:80]: Rename to `add_create_private...`

File: ca/django_ca/backends/base.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        :py:attr:`~django_ca.backends.base.KeyBackend.title`.
+
+        Return ``None`` if you don't need to create such a group.
+        """
+        return parser.add_argument_group(
+            f"{self.alias} key storage",
+            f"Arguments for using private keys stored with the {self.alias} backend.",
+        )
+
+    def add_private_key_arguments(self, group: ArgumentGroup) -> None:  # pylint: disable=unused-argument
```
Comment (mathiasertl) [inline ca/django_ca/backends/base.py:121]: Rename to `add_create_private...`

File: ca/django_ca/backends/base.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        self, key_type: ParsableKeyType, options: Dict[str, Any]
+    ) -> CreatePrivateKeyOptionsTypeVar:
+        """Load options to create private keys into a Pydantic model.
+
+        `options` is the dictionary of arguments to ``manage.py init_ca`` (including default values). The
+        returned model will be passed to :py:func:`~django_ca.backends.base.KeyBackend.create_private_key`.
+        """
+
+    @abc.abstractmethod
+    def get_load_private_key_options(self, options: Dict[str, Any]) -> UsePrivateKeyOptionsTypeVar:
```
Comment (mathiasertl) [inline ca/django_ca/backends/base.py:158]: Rename to `get_use_...`

File: ca/django_ca/backends/base.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        `options` is the dictionary of arguments to ``manage.py init_ca`` (including default values). The key
+        backend is expected to be able to sign certificates and CRLs using the options provided here.
+        """
+
+    @abc.abstractmethod
+    def get_store_private_key_options(self, options: Dict[str, Any]) -> StorePrivateKeyOptionsTypeVar:
+        """Get options used when storing private keys."""
+
+    @abc.abstractmethod
+    def get_load_parent_private_key_options(self, options: Dict[str, Any]) -> UsePrivateKeyOptionsTypeVar:
```
Comment (mathiasertl) [inline ca/django_ca/backends/base.py:170]: Rename to `get_use_...`

File: ca/django_ca/api/schemas.py [hunk 8]
```diff
@@ -143,7 +143,7 @@ class Meta(X509BaseSchema.Meta):  # pylint: disable=missing-class-docstring
     @staticmethod
     def resolve_can_sign_certificates(obj: CertificateAuthority) -> bool:
         """Resolve the can_sign_certificates flag."""
-        return obj.key_exists
+        return obj.is_usable(options=None)
```
Comment (mathiasertl) [inline ca/django_ca/api/schemas.py:146]: Remove options here

File: ca/django_ca/admin.py [hunk 9]
```diff
@@ -677,10 +674,7 @@ def get_ca_details(self) -> Dict[int, Dict[str, Any]]:
 
     def has_add_permission(self, request: HttpRequest) -> bool:
         # Only grant add permissions if there is at least one usable CA
-        for ca in CertificateAuthority.objects.usable():
-            if ca.key_exists:
-                return True
-        return False
+        return CertificateAuthority.objects.usable().exists()
```
Comment (mathiasertl) [inline ca/django_ca/admin.py:677]: Restore

File: ca/django_ca/admin.py [hunk 10]
```diff
@@ -659,9 +659,6 @@ def get_ca_details(self) -> Dict[int, Dict[str, Any]]:
         """Get CA details for the embedded JSON data."""
         data: Dict[int, Dict[str, Any]] = {}
         for ca in CertificateAuthority.objects.usable():
-            if ca.key_exists is False:
-                continue
-
```
Comment (mathiasertl) [inline ca/django_ca/admin.py:664]: Restore

File: ca/django_ca/management/actions.py [hunk 11]
```diff
@@ -241,6 +237,17 @@ def parse_value(self, value: int) -> int:
...
 
 
+class KeyBackendAction(SingleValueAction[str, KeyBackend]):
+    def __init__(self, **kwargs: Any) -> None:
+        kwargs.setdefault("choices", list(ca_settings.CA_KEY_BACKENDS))
+        kwargs.setdefault("default", key_backends[ca_settings.CA_DEFAULT_KEY_BACKEND])
+        super().__init__(**kwargs)
+
+    def parse_value(self, value: str) -> KeyBackend:
+        print("parse_value", value, key_backends[value])
```
Comment (mathiasertl) [inline ca/django_ca/management/actions.py:247]: Remove print

File: ca/django_ca/ca_settings.py (hunk 3 above)
Comment (mathiasertl) [reply ca/django_ca/ca_settings.py]: will be done in the docs

**Pull request merged**